src/agent/internals/src/main/java/com/google/devtools/cdbg/debuglets/java/GcpEnvironment.java

```diff
@@ -173,6 +175,36 @@ static Map<String, String> getDebuggeeLabels() {
     return labels;
   }
 
+  /**
+   * Returns the user dir of a Java8 App Engine Standard instance. This is the base directory where
+   * the user's files are deployed. If no Java 8 App Engine Standard user dir could be determined
+   * null is returned.  It is safe to call this regardless of environment, as long as the return
+   * value is checked.
+   *
+   * <p> While there is a system property defined, "user.dir", in Java8 App Engine environments, it
+   * is unfortunately not set at the point in time this agent code runs, so it cannot be relied
+   * upon. Here we use the heuristic that the WEB-INF directory is deployed at the root of the user
+   * dir, and the deployed agent is deployed somewhere under the user dir. This means we can search
+   * through the agent directory's ancestors looking for the WEB-INFO directory, and once found
+   * we've found the user dir.
+   *
+   * @return the Java 8 App Engine Standard user directory if it was able to be determined, null
+   * otherwise.
+   */
+  public static String getAppEngineJava8UserDir(String agentDir) {
+    File currentDir = new File(agentDir);
+
+    while (currentDir != null) {
+      if (new File(currentDir.getAbsolutePath(), "WEB-INF").exists()) {
```

Comment:
Isn't this going to trigger for other environments other than GAE Java8?  eg. GCE Tomcat?

Reply:
I was going on the assumption that hitting a false positive wasn't harmful, as it would just increase the chance of finding the class files to index. In the case of jetty/tomcat on GCE, in those cases I believe usually the WEB-INF directory is found under "webapps/ROOT/WEB-INF", in the java8 case here WEB-INF is placed at the root of the user dir witthout the webapps/ROOT prefix.

That said I'm ok with adding an extra guard,  we can check if the GAE_AGENTPATH_OPTS environment variable is set, this will increase the likelihood we're on java 8, since that's the only environment it would be expected to be set.

Reply:
I'd prefer checking that environment variable

Reply:
Done, though in the end I went with checking the GAE_RUNTIME environment variable for the value of 'java8'.